include/SolverOptions.hpp

```diff
@@ -0,0 +1,63 @@
+#ifndef SOLVER_OPTIONS_HPP
+#define SOLVER_OPTIONS_HPP
+
+// An enumeration of possible preconditioners for the problem.
+enum preconditioner_id { BLOCK_DIAGONAL, SIMPLE, ASIMPLE, YOSHIDA };
```

Comment:
In modern C++, it is generally better to use an `enum class` rather than a simple `enum`:
```suggestion
enum class preconditioner_id {
  BLOCK_DIAGONAL,
  SIMPLE,
  ASIMPLE,
  YOSHIDA
};
```
The `enum class` offers a good deal of semantic advantages with respect to the `enum` (e.g. it is not implicitly casted to the underlying integer type).

(on a minor note, the preconditioner is actually called "Yosida").